File: Java/Civitas/src/civitas/CivitasJuego.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Arrays;


public class CivitasJuego{
	private int indiceJugadorActual;
	private ArrayList<Jugador> jugadores;
	private GestorEstados gestorEstados;
	private MazoSorpresas mazo;
	private Tablero tablero;
	private EstadosJuego estado;
        
        
	private void avanzaJugador(){
		Jugador jugadorActual=jugadores.get(indiceJugadorActual);
		int posicionActual=jugadorActual.getNumCasillaActual();
		int tirada=Dado.getInstance().tirar();
		int posicionNueva=tablero.nuevaPosicion(posicionActual,tirada);
		Casilla_Descanso casilla=tablero.getCasilla(posicionNueva);
		contabilizarPasosPorSalida(jugadorActual);
                jugadorActual.moverACasilla(posicionNueva);
		casilla.recibeJugador(indiceJugadorActual, jugadores);
		contabilizarPasosPorSalida(jugadorActual); 
	}

	public CivitasJuego(ArrayList <String> nombres){
            
                //Debug mode of Dado.
                //Dado.getInstance().setDebug(true);
            
		jugadores=new ArrayList<>();
		for(int i=0;i<nombres.size();i++){
			jugadores.add(new Jugador(nombres.get(i)));
		}
		gestorEstados=new GestorEstados();
		estado=gestorEstados.estadoInicial();
		indiceJugadorActual=Dado.getInstance().quienEmpieza(nombres.size());
		mazo=new MazoSorpresas();
		inicializarTablero(mazo);
		inicializarMazoSorpresas(tablero);
	}

	public boolean cancelarHipoteca(int ip){
		return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
	}

	public boolean comprar(){
		Jugador jugador = jugadores.get(indiceJugadorActual);
		int num = jugador.getNumCasillaActual();
		TituloPropiedad titulo = ((Casilla_Propiedad)tablero.getCasilla(num)).getTituloPropiedad();
		return jugador.comprar(titulo);
	}

	public boolean construirCasa(int ip){
		return jugadores.get(indiceJugadorActual).construirCasa(ip);
	}

	public boolean construirHotel(int ip){
		return jugadores.get(indiceJugadorActual).construirHotel(ip);
	}

	
	private void contabilizarPasosPorSalida(Jugador jugadorActual){
		while(tablero.getPorSalida()>0){
			jugadorActual.pasaPorSalida();
		}
	}

	public boolean finalDelJuego(){
		boolean result=false;
		for(int i=0;i<jugadores.size();i++){
			if(jugadores.get(i).enBancarrota()){
				result=true;
			}
		}
		return result;
	}

	public Casilla_Descanso getCasillaActual(){
		return tablero.getCasilla(jugadores.get(indiceJugadorActual).getNumCasillaActual());
	}

	public Jugador getJugadorActual(){
		return jugadores.get(indiceJugadorActual);
	}

	public boolean hipotecar(int ip){
		return jugadores.get(indiceJugadorActual).hipotecar(ip);
	}

	private void inicializarMazoSorpresas(Tablero tablero){
		mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, 500, "Concurso de belleza, cobra 500."));
		mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, -500, "Multa de trafico, paga 500."));
		mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, tablero.getCarcel(), "Visita a tu compañero de la carcel"));
		mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 2, "Ruleta rusa."));
		mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 3, "Ruleta rusa."));
		mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, 50, "Dineros por cada casa"));
		mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, -50, "Dineros (negativos) por casas"));
		mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo));
		mazo.alMazo(new Sorpresa(TipoSorpresa.IRCARCEL, tablero, 0, "Sorpresa! Carcel"));
	}

	private void inicializarTablero(MazoSorpresas mazo){
		tablero = new Tablero(5);
		float revalorizacion = (float) 1.15;
		// =>  La casilla del inicio se introduce automáticamente
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Calle Caramelo", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Plaza Galleta", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Sorpresa(mazo, "Esquina de las Pizzas"));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Estación de los gofres", 50, revalorizacion, 150, 200, 120)
		));
		// =>  Aquí meterá tablero a la carcel
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Ruta glaseada", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Rio Cremoso", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Sorpresa(mazo, "Puente Hamburguesa"));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Zoo de los dulces", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Descanso("Parking Gratuito"));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Cruce Gominolas", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Chocolate Street", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Impuesto(300, "Impuesto del Azucar"));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Avenida de las Avellanas", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeJuez();
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Banco de monedas de chocolate", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Palacio del dulce de leche", 50, revalorizacion, 150, 200, 120)
		));
		tablero.añadeCasilla(new Casilla_Sorpresa(mazo, "Puente Hamburguesa"));
		tablero.añadeCasilla(new Casilla_Propiedad(
		  new TituloPropiedad("Hotel de los batidos", 50, revalorizacion, 150, 200, 120)
		));
	}

	private void pasarTurno(){
		indiceJugadorActual = (indiceJugadorActual + 1) % jugadores.size();
	}

	//Modified from private to public to match with the script.

	public ArrayList<Jugador> ranking(){
		int p;
		ArrayList<Jugador> copia;
		copia=new ArrayList<>(jugadores);
		Collections.sort(copia);
		return copia;
	}

	public Boolean salirCarcelPagando(){
		return jugadores.get(indiceJugadorActual).salirCarcelPagando();
	}

	public Boolean salirCarcelTirando(){
		return jugadores.get(indiceJugadorActual).salirCarcelTirando();
	}

	public OperacionesJuego siguientePaso(){
		Jugador jugadorActual=jugadores.get(indiceJugadorActual);
		OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual,estado);
		contabilizarPasosPorSalida(jugadorActual);
		if(operacion==OperacionesJuego.PASAR_TURNO){
			this.pasarTurno();
			this.siguientePasoCompletado(operacion);
		}
		else if(operacion==OperacionesJuego.AVANZAR){
			this.avanzaJugador();
			this.siguientePasoCompletado(operacion);
		}
		contabilizarPasosPorSalida(jugadorActual);
                return operacion;
	}
	
	public void siguientePasoCompletado(OperacionesJuego operacion){
			estado=gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual),estado,operacion);
	}

	public boolean vender(int ip){
		return jugadores.get(indiceJugadorActual).vender(ip);
	}
	

	
}

File: Ruby/CivitasR/lib/CivitasJuego.rb
#encoding:utf-8

require_relative "Tablero"
require_relative "Diario"
require_relative "Dado"
require_relative "MazoSorpresas"
require_relative "Jugador"
require_relative "Enum"
require_relative "GestorEstados"

module Civitas
  class CivitasJuego
    
    def initialize(nombres)
      @jugadores = nombres.map { |nombre|  Jugador.new(nombre) }
      @gestorEstados = Gestor_estados.new
      @estado = @gestorEstados.estado_inicial
      @indiceJugadorActual = Dado.instance.quienEmpieza(nombres.size)
      @mazo = MazoSorpresas.new(true)
      inicializaTablero(@mazo)
      inicializaMazoSorpresas(@tablero)
    end

    def inicializaTablero(mazo)
      @tablero = Tablero.new(5)
      revalorizacion = 1.15
      # =>  La casilla del inicio se introduce automáticamente
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Calle Caramelo", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Plaza Galleta", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Sorpresa.new(mazo, "Esquina de las Pizzas"))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Estación de los gofres", 50, revalorizacion, 150, 200, 120)
      ))
      # =>  Aquí meterá tablero a la carcel
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Ruta glaseada", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Rio Cremoso", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Sorpresa.new(mazo, "Puente Hamburguesa"))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Zoo de los dulces", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Descanso.new("Parking Gratuito"))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Cruce Gominolas", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Chocolate Street", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Impuesto.new(300, "Impuesto del Azucar"))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Avenida de las Avellanas", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeJuez
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Banco de monedas de chocolate", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Palacio del dulce de leche", 50, revalorizacion, 150, 200, 120)
      ))
      @tablero.añadeCasilla(Casilla_Sorpresa.new(mazo, "Puente Hamburguesa"))
      @tablero.añadeCasilla(Casilla_Calle.new(
          TituloPropiedad.new("Hotel de los batidos", 50, revalorizacion, 150, 200, 120)
      ))
    end
    private :inicializaTablero

    def inicializaMazoSorpresas(tablero)
      @mazo.alMazo(Sorpresa_IrCarcel.new("Carcel",tablero))
      @mazo.alMazo(Sorpresa_SalirCarcel.new("Carta de salir de la carcel",@mazo))
      @mazo.alMazo(Sorpresa_MoverACasilla.new("Te mudas",tablero,10))
      @mazo.alMazo(Sorpresa_MoverACasilla.new("A la casilla de salida",tablero,1))
      @mazo.alMazo(Sorpresa_MoverACasilla.new("Muevete de casilla",tablero,20))
      @mazo.alMazo(Sorpresa_PorCasaHotel.new("Reparaciones",1))
      @mazo.alMazo(Sorpresa_PorCasaHotel.new("Tuberias oxidadas",1))
      @mazo.alMazo(Sorpresa_PagarCobrar.new("Pagas la comunidad",-100))
      @mazo.alMazo(Sorpresa_PagarCobrar.new("Concurso de belleza",100))
    end
    private :inicializaMazoSorpresas

    def contabilizarPasosPorSalida(jugadorActual)
      while @tablero.getPorSalida != 0
        jugadorActual.pasaPorSalida
      end
    end
    private :contabilizarPasosPorSalida

		def pasarTurno
			@indiceJugadorActual = (@indiceJugadorActual + 1) % @jugadores.size
		end
    private :pasarTurno

		def siguientePasoCompletado(operacion)
			@estado = @gestorEstados.siguiente_estado(@jugadores[@indiceJugadorActual], @estado, operacion)
		end

		def construirCasa(ip)
			return @jugadores[@indiceJugadorActual].construirCasa(ip)
		end

		def construirHotel(ip)
			return @jugadores[@indiceJugadorActual].construirHotel(ip)
		end

		def vender(ip)
			return @jugadores[@indiceJugadorActual].vender(ip)
		end

		def hipotecar(ip)
			return @jugadores[@indiceJugadorActual].hipotecar(ip)
		end

		def cancelarHipoteca(ip)
			return @jugadores[@indiceJugadorActual].cancelarHipoteca(ip)
		end

		def salirCarcelPagando
			return @jugadores[@indiceJugadorActual].salirCarcelPagando
		end

		def salirCarcelTirando
			return @jugadores[@indiceJugadorActual].salirCarcelTirando
		end

		def finalDelJuego
			return @jugadores.any? { |jugador| jugador.enBancarrota() }
		end

    # => Ranking se hace público para que pueda ser mostrado si el juego se acaba
		def ranking
			return @jugadores.sort!.reverse!
		end
    
    def getJugadorActual
      return @jugadores[@indiceJugadorActual]
    end
    
    def getCasillaActual
      return @tablero.getCasilla(@jugadores[@indiceJugadorActual].numCasillaActual)
    end
    
    def avanzaJugador
      jugadorActual = @jugadores[@indiceJugadorActual] #1.1
      posicionNueva = @tablero.nuevaPosicion( jugadorActual.numCasillaActual,Dado.instance.tirar )  #1.2, 1.3 y 1.4
      casilla = @tablero.getCasilla(posicionNueva) #1.5
      contabilizarPasosPorSalida(jugadorActual) #1.6
      jugadorActual.moverACasilla(posicionNueva) #1.7
      casilla.recibeJugador(@indiceJugadorActual,@jugadores) #1.8
      contabilizarPasosPorSalida(jugadorActual) #1.9
    end
    private :avanzaJugador

    def siguientePaso
      # 1, 2
      contabilizarPasosPorSalida(@jugadorActual)
      operacion = @gestorEstados.operaciones_permitidas(@jugadores[@indiceJugadorActual],@estado);
      case operacion
      when OperacionesJuego::PASAR_TURNO
        pasarTurno # 3
        siguientePasoCompletado(operacion) # 4
      when OperacionesJuego::AVANZAR
        avanzaJugador # 5
        siguientePasoCompletado(operacion) # 6
      
      end
      contabilizarPasosPorSalida(@jugadorActual)
      return operacion
    end

    def comprar
      jugadorActual = @jugadores[@indiceJugadorActual] #1
      numCasillaActu = jugadorActual.numCasillaActual  #2
      casilla = @tablero.getCasilla(numCasillaActu) #3
      titulo = casilla.titulo #4
      res = jugadorActual.comprar(titulo) #5
      return res
    end

    def infoJugadorTexto
      jugador = @jugadores[ @indiceJugadorActual ]
      s = jugador.toString + "\n" +
        "Propiedades:  \n"
      jugador.nombrePropiedades.each { |p| s += p + '\n' }
      s += "\n" + @tablero.getCasilla( jugador.numCasillaActual ).toString
      return s
    end
    
    def actualizarInfo
      puts @jugadores[@indiceJugadorActual].toString
      puts "Propiedades del jugador: "
      @jugadores[@indiceJugadorActual].getPropiedades.each do |prop|
        puts prop.toString
      end
      puts "Casilla Actual: "
      puts @tablero.getCasilla(@jugadores[@indiceJugadorActual].numCasillaActual).toString
      if @jugadores.any? { |jugador| jugador.enBancarrota() }
        puts "Ranking: "
        for i in 1..@jugadores.size
          puts i.to_s+". "+@jugadores[i].getNombre+" - "+@jugadores[i].getSaldo.to_s+"€ "
        end
      end
    end

  end
end
